instrumentation/ruby_kafka/lib/opentelemetry/instrumentation/ruby_kafka/patches/consumer.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module RubyKafka
+      module Patches
+        # The Consumer module contains the instrumentation patch for the Consumer class
+        module Consumer
+          def each_message(min_bytes: 1, max_bytes: 10_485_760, max_wait_time: 1, automatically_mark_as_processed: true)
+            super do |message|
+              attributes = {
+                'messaging.system' => 'kafka',
+                'messaging.destination' => message.topic,
+                'messaging.destination_kind' => 'topic',
+                'messaging.kafka.partition' => message.partition,
+                'messaging.kafka.offset' => message.offset
+              }
+
+              attributes['messaging.kafka.message_key'] = message.key if message.key
+
+              link = OpenTelemetry::Trace::Link.new(
+                OpenTelemetry.propagation.text.extract(message.headers)
+              )
+
+              tracer.in_span("#{message.topic} process", links: [link], attributes: attributes, kind: :consumer) do
+                yield message
+              end
+            end
+          end
+
+          def each_batch(min_bytes: 1, max_bytes: 10_485_760, max_wait_time: 1, automatically_mark_as_processed: true)
+            super do |batch|
+              attributes = {
+                'messaging.system' => 'kafka',
+                'messaging.destination' => batch.topic,
+                'messaging.destination_kind' => 'topic',
+                'messaging.kafka.partition' => batch.partition,
+                'messaging.kafka.offset_lag' => batch.offset_lag,
+                'messaging.kafka.highwater_mark_offset' => batch.highwater_mark_offset,
+                'messaging.kafka.message_count' => batch.messages.count
+              }
+
+              links = batch.messages.map do |message|
+                OpenTelemetry::Trace::Link.new(
+                  OpenTelemetry.propagation.text.extract(message.headers)
+                )
+              end
```

Comment:
I'm not sure how the tests are working 🤔 . The `Link` initializer takes a `SpanContext`, but the propagator's `extract` returns a `Context` that *might* contain a `Span` that wraps a `SpanContext`. We need to dig through an unfortunate number of layers to get to a `SpanContext` that might not be there.
```suggestion
              links = batch.messages.map do |message|
                span_context = OpenTelemetry::Trace.current_span(OpenTelemetry.propagation.text.extract(message.headers)).context
                OpenTelemetry::Trace::Link.new(span_context) if span_context.valid?
              end
              links.compact!
```

Reply:
My tests might be doing the wrong thing, for more isolated view on what I did with links this commit doesn't have the extra noise. https://github.com/open-telemetry/opentelemetry-ruby/pull/476/commits/8a1ef1b15af53b1e636e4b35f01a84706f23f237